pkg/runtime/sequencer/cache.go

```diff
@@ -15,15 +15,225 @@ package sequencer
 
 import (
 	"context"
+	"errors"
 	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+	"mosn.io/pkg/utils"
+	"sync"
+	"sync/atomic"
+	"time"
 )
 
+const defaultSize = 10000
+const defaultLimit = 1000
+const defaultRetry = 5
+const waitTime = time.Second
+
+// DoubleBuffer is double segment id buffer.
+// There are two buffers in DoubleBuffer: InUseBuffer is in use, BackUpBuffer is a backup buffer.
+// Their default capacity is 1000. When the InUseBuffer usage exceeds 30%, the BackUpBuffer will be initialized.
+// When InUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key              string
+	Size             int
+	InUseBuffer      *Buffer
+	BackUpBufferChan chan *Buffer
+	// 1 means getting BackUpBuffer，0 means not
+	Processing uint32
+	lock       sync.Mutex
+	Store      sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:              key,
+		Size:             defaultSize,
+		Store:            store,
+		BackUpBufferChan: make(chan *Buffer, 1),
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		err := d.swap()
+		if err != nil {
+			return 0, err
+		}
+	}
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+			//one case: here and swap are performed simultaneously,
+			//swap fast,no check will cover old BackUpBuffer
+
+			//remove lock,add channel&CAS for visibility
+			//check  not on processing and bufChannel  nil
+			if atomic.CompareAndSwapUint32(&d.Processing, 0, 1) {
+				defer atomic.StoreUint32(&d.Processing, 0)
+				if len(d.BackUpBufferChan) != 0 {
```

Comment:
how about retry until `getNewBuffer` success?
so in swap, you can just wait on BackUpBufferChan or timeout.

Reply:
@MoonShining   Gooood idea, put retry here and can  remove processing status